text/0000-enable-embroider.md

```diff
@@ -0,0 +1,190 @@
+
+## Detailed design
+
+This design will cover a few topics. The first discusses the explicit file changes that happen for your application to use Embroider. Next it will cover how both new applications move to Embroider and how existing applications will migrate to it. Lastly, it will cover some deprecated concepts that no longer apply under Embroider along with how we get community addon adoption.
+
+### Packages
+
+The following packages will be added to package.json:
+
+- @embroider/compat - Stage 1 is responsible for upgrading addons from their current implmentations ("V1") into a new format called "V2 Packages". This package also contains the entry point method that is invoked in your ember-cli-build.js file (shown below).
+- @embroider/core - Stage 2 is responsible for taking a collection of V2 packages and abstracting away the "emberness" into standard Javascript which all packagers can understand.
+- @embroider/webpack - Stage 3 of Embroider's design allows you to bring your own packager. By default Embroider ships with and recommends using Webpack
+
+### Build File Changes:
+
+Embroider replaces the logic provided by `ember-app.js`. By default it runs in a "safe" mode that provides the best backwards compatibility guarantees but prevents some optimizations from happening. This safe mode configuration is what we will ship by default:
+
+```diff
+-return app.toTree();
++const { Webpack } = require('@embroider/webpack');
++return require('@embroider/compat').compatBuild(app, Webpack);
+```
+
+### New application flow
+
+The above changes will be applied to the blueprint for `ember new`. Additionally, the build flag option of `--embroider` will exist so that the option will be recorded in `config/ember-cli-update.json` like:
+
+```json
+{
+  "schemaVersion": "1.0.0",
+  "packages": [{
+    "name": "ember-cli",
+    "version": "X.X.X",
+    "blueprints": [{
+      "name": "app",
+      "outputRepo": "https://github.com/ember-cli/ember-new-output",
+      "codemodsSource": "ember-app-codemods-manifest@1",
+      "isBaseBlueprint": true,
+      "options": [
+        "--embroider"
+      ]
+    }]
+  }]
+}
+```
+
+This will ensure that `ember-cli-update` only applies these changes to applications that are already on Embroider.
+
+### Existing application migration flow
+
+In order to make the migration as simple as possible, a codemod will be created. The codemod CLI will provide two commands: `preflight` and `migrate`.
+
+Example usage:
+
+`npx @embroider/cli preflight`
+`npx @embroider/cli migrate`
+
+#### Preflight
+
+The goal of this command is to give users an easy yes or no answer to the question: can I start migrating to Embroider? As an example, we can check if the project is on the minimal set of package versions of known "fixed" addons along with other static checks.
+
+#### Migrate
+
+The migrate command is purely to codemod the file changes above to get your app onto Embroider.
+
+### Deprecations
```

Comment:
I believe some of the existing Ember CLI build hooks used by addons are no longer supported in Embroider land. Can we be more explicit about what these hooks are, and if possible some guidance on how to migrate away from using them?

Should we also have official deprecations for these unsupported hooks?